pkg/auth/kubesecret.go

```diff
@@ -0,0 +1,182 @@
+func (kubelistener *KubeSecretListener) deleteDockerConfig(key string) error {
+	kubelistener.configMu.Lock()
+	delete(kubelistener.dockerConfigs, key)
+	kubelistener.configMu.Unlock()
+	_, exists := kubelistener.dockerConfigs[key]
+	if exists {
+		return fmt.Errorf("failed to delete docker config")
+	}
+	return nil
+}
+
+func (kubelistener *KubeSecretListener) SyncKubeSecrets(ctx context.Context, clientset *kubernetes.Clientset) error {
+	if kubelistener.informer != nil {
+		return nil
+	}
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = "type=" + string(corev1.SecretTypeDockerConfigJson)
+				return clientset.CoreV1().Secrets(metav1.NamespaceAll).List(context.Background(), options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = "type=" + string(corev1.SecretTypeDockerConfigJson)
+				return clientset.CoreV1().Secrets(metav1.NamespaceAll).Watch(context.Background(), options)
+			}},
+		&corev1.Secret{},
+		0,
+		cache.Indexers{},
+	)
+	kubelistener.informer = informer
+	kubelistener.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				logrus.WithError(err).Infof("failed to get key for secret from cache")
+				return
+			}
+			if err := kubelistener.addDockerConfig(key, obj); err != nil {
+				logrus.WithError(err).Errorf("failed to add a new dockerconfigjson")
+				return
+			}
+		},
+		UpdateFunc: func(old, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err != nil {
+				logrus.WithError(err).Infof("failed to get key for secret from cache")
+				return
+			}
+			if err := kubelistener.addDockerConfig(key, new); err != nil {
+				logrus.WithError(err).Errorf("failed to add a new dockerconfigjson")
+				return
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				logrus.WithError(err).Infof("failed to get key for secret from cache")
+			}
+			if err := kubelistener.deleteDockerConfig(key); err != nil {
+				logrus.WithError(err).Errorf("failed to delete docker config")
+				return
+			}
+		}},
+	)
+	go kubelistener.informer.Run(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+		return fmt.Errorf("timed out for syncing cache")
+	}
+	return nil
+}
+
+func (kubelistener *KubeSecretListener) GetCredentialsStore(host string) *PassKeyChain {
+	kubelistener.configMu.Lock()
+	defer kubelistener.configMu.Unlock()
+	for _, dockerConfig := range kubelistener.dockerConfigs {
+		// Find the auth for the host.
+		authConfig, err := dockerConfig.GetAuthConfig(host)
+		if err != nil {
+			logrus.WithError(err).Infof("failed to get auth config for host %s", host)
+			continue
+		}
+		if len(authConfig.Username) != 0 && len(authConfig.Password) != 0 {
```

Comment:
Is it legal that the password is empty? 

```
username: "tom",
password: "",
```

Reply:
--docker-username, --docker-password and --docker-server is required